File: lecode684.py
from typing import List
class Solution:
    # 暴力
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        father = [x for x in range(n + 1)]
        def find(x :int):
            if father[x] == x:
                return x
            father[x] = find(father[x])
            return father[x]
        def union(m : int,n : int):
            m = find(m)
            n = find(n)
            if m == n :
                return
            elif m < n :
                father[n] = m
            else :
                father[m] = n
        index = -1
        res = []
        for i in range(n):
            for idx,val in enumerate(edges):
                if i == idx :
                    continue
                union(val[0],val[1])
            temp = 0
            for j in range(1,n + 1):
                if father[j] == j :
                    temp += 1
            if temp == 1:
                res.append(edges[i])
            father = [x for x in range(n + 1)]
        return res[-1]

File: lecode684.cpp
#include <iostream>
#include <vector>
using namespace std;
// 非暴力
class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size();
        for(int i = 0;i < n + 1;i ++) father.emplace_back(i);
        for (int i = 1;i < this->father.size();i++){
            if(union_(edges[i][0],edges[i][1])){
                return edges[i];
            }
        }
        return {};
    }
    int find(int x){
        if(father[x] == x){
            return x;
        }
        return father[x] = find(father[x]);
    }
    bool union_(int a,int b){
        a = find(a);
        b = find(b);
        if(a == b){
            return false;
        }else if (a < b){
            father[b] = a;
        }else{
            father[a] = b;
        }
        return true;
    }
private:
    vector<int> father;
};
